src/main/java/hudson/tasks/SMTPAuthentication.java

```diff
@@ -1,40 +1,131 @@
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.AncestorInPath;
 import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.util.Collections;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.StreamSupport;
 
 /**
  * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
  */
 public class SMTPAuthentication extends AbstractDescribableImpl<SMTPAuthentication> {
 
-    private String username;
+    private static final Logger LOGGER = Logger.getLogger(SMTPAuthentication.class.getName());
+
+    /** use StandardUsernamePasswordCredentials instead */
+    @Deprecated
+    private transient String username;
+
+    /** use StandardUsernamePasswordCredentials instead */
+    @Deprecated
+    private transient Secret password;
 
-    private Secret password;
+    /** The ID of the Jenkins Username/Password credential to use. */
+    private String credentialsId;
 
     @DataBoundConstructor
-    public SMTPAuthentication(String username, Secret password) {
+    public SMTPAuthentication(String credentialsId) {
+        this.credentialsId = credentialsId;
+    }
+
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    SMTPAuthentication(String username, Secret password) {
         this.username = username;
         this.password = password;
+        readResolve();
     }
 
-    public String getUsername() {
-        return username;
+    public String getCredentialsId() {
+        return credentialsId;
     }
 
-    public Secret getPassword() {
-        return password;
+    private Object readResolve() {
+        if (username != null || password != null) {
+            LOGGER.log(Level.CONFIG, "Migrating the Mailer SMTP authentication details to credential...");
+
+            final CredentialsStore store = StreamSupport.stream(CredentialsProvider.lookupStores(Jenkins.get()).spliterator(), false)
+                    .filter(s -> s instanceof SystemCredentialsProvider.StoreImpl)
+                    .findFirst()
+                    .orElseThrow(() -> new RuntimeException("Could not migrate the Mailer SMTP authentication details to credential, as the system credentials provider was missing."));
+
+            final boolean isSuccess = Retryable.retry(9, (attempt) -> {
+                LOGGER.log(Level.CONFIG, "Attempt {0}...", attempt);
+
+                final String id = UUID.randomUUID().toString();
+
+                final StandardUsernamePasswordCredentials migratedCredential = new UsernamePasswordCredentialsImpl(
+                        CredentialsScope.GLOBAL,
+                        id,
+                        "Mailer SMTP authentication credentials (migrated)",
+                        username,
+                        password.getEncryptedValue());
+
+                store.addCredentials(Domain.global(), migratedCredential);
```

Comment:
I ignored the return value, as a failure should be expressed by the method throwing an exception. Is this alright?